### PR TITLE
[mlir] Fix a zero stride canonicalizer crash

### LINKED ISSUE #73383: [mlir] Canonicalizer crashed with assertion failure.
git version: 024718313b526a4530d7cd58886fcc157247e0ff

system: Ubuntu 20.04.6 LTS (Focal Fossa)

reproduced with: ```mlir-opt --canonicalize a.mlir```

a.mlir:
```
#map = affine_map<(d0)[s0] -> (-((d0 * 2) floordiv 16))>
func.func @func1(%arg0: i1, %arg1: index) {
  %c0 = arith.constant 0 : index
  %c1 = arith.constant 1 : index
  %c8 = arith.constant 8 : index
  %c23 = arith.constant 23 : index
  %alloc_8 = memref.alloc() {alignment = 64 : i64} : memref<2xi64>
  %22 = memref.rank %alloc_8 : memref<2xi64>
  %44 = llvm.mlir.constant(0 : i64) : i64
  %45 = builtin.unrealized_conversion_cast %44 : i64 to index
  %54 = affine.max #map(%22)[%c23]
  %alloc_46 = memref.alloc() {alignment = 64 : i64} : memref<1x1x1xi16>
  %80 = bufferization.to_tensor %alloc_46 : memref<1x1x1xi16>
  %81 = builtin.unrealized_conversion_cast %80 : tensor<1x1x1xi16> to !spirv.array<1 x i32>
  %alloc_47 = memref.alloc() {alignment = 64 : i64} : memref<1x1x1xi16>
  %alloc_48 = memref.alloc() {alignment = 64 : i64} : memref<1x1xi32>
  %82 = builtin.unrealized_conversion_cast %c0 : index to i32
  spirv.Branch ^bb1(%82, %81 : i32, !spirv.array<1 x i32>)
^bb1(%83: i32, %84: !spirv.array<1 x i32>):  // 2 preds: ^bb0, ^bb5
  %85 = builtin.unrealized_conversion_cast %83 : i32 to index
  spirv.Branch ^bb3(%82, %84 : i32, !spirv.array<1 x i32>)
^bb3(%88: i32, %89: !spirv.array<1 x i32>):  // 2 preds: ^bb2, ^bb4
  %90 = builtin.unrealized_conversion_cast %88 : i32 to index
  %94 = memref.load %alloc_48[%85, %90] : memref<1x1xi32>
  %95 = arith.index_cast %94 : i32 to index
  %subview = memref.subview %alloc_47[%85, %c8, %c0] [%c1, %c1, %c1] [%c1, %c1, %c1] : memref<1x1x1xi16> to memref<?x?x?xi16, strided<[?, ?, ?], offset: ?>>
  %alloc_49 = memref.alloc() {alignment = 64 : i64} : memref<1x1x1xi16>
  %subview_50 = memref.subview %alloc_49[%85, %95, %c0] [%95, %c1, %c1] [%c1, %54, %45] : memref<1x1x1xi16> to memref<?x?x?xi16, strided<[?, ?, ?], offset: ?>>
  memref.copy %subview, %subview_50 : memref<?x?x?xi16, strided<[?, ?, ?], offset: ?>> to memref<?x?x?xi16, strided<[?, ?, ?], offset: ?>>
  return
}
```

trace:
```
<unknown>:0: error: strides must not be zero
mlir-opt: /data/bin/llvm-project/mlir/include/mlir/IR/StorageUniquerSupport.h:180: static ConcreteT mlir::detail::StorageUserBase<ConcreteT, BaseT, StorageT, UniquerT, Traits>::get(mlir::MLIRContext*, Args&& ...) [with Args = {long int, llvm::ArrayRef<long int>}; ConcreteT = mlir::StridedLayoutAttr; BaseT = mlir::Attribute; StorageT = mlir::detail::StridedLayoutAttrStorage; UniquerT = mlir::detail::AttributeUniquer; Traits = {mlir::MemRefLayoutAttrInterface::Trait}]: Assertion `succeeded(ConcreteT::verify(getDefaultDiagnosticEmitFn(ctx), args...))' failed.
PLEASE submit a bug report to https://github.com/llvm/llvm-project/issues/ and include the crash backtrace.
Stack dump:
0.      Program arguments: /data/bin/llvm-project/build/bin/mlir-opt --inline a.mlir
 #0 0x000055b03dcbd06f llvm::sys::PrintStackTrace(llvm::raw_ostream&, int) (/data/bin/llvm-project/build/bin/mlir-opt+0x160306f)
 #1 0x000055b03dcba0c4 SignalHandler(int) Signals.cpp:0:0
 #2 0x00007f31f4063420 __restore_rt (/lib/x86_64-linux-gnu/libpthread.so.0+0x14420)
 #3 0x00007f31f3b4e00b raise (/lib/x86_64-linux-gnu/libc.so.6+0x4300b)
 #4 0x00007f31f3b2d859 abort (/lib/x86_64-linux-gnu/libc.so.6+0x22859)
 #5 0x00007f31f3b2d729 (/lib/x86_64-linux-gnu/libc.so.6+0x22729)
 #6 0x00007f31f3b3efd6 (/lib/x86_64-linux-gnu/libc.so.6+0x33fd6)
 #7 0x000055b040a1c2cb mlir::StridedLayoutAttr mlir::detail::StorageUserBase<mlir::StridedLayoutAttr, mlir::Attribute, mlir::detail::StridedLayoutAttrStorage, mlir::detail::AttributeUniquer, mlir::MemRefLayoutAttrInterface::Trait>::get<long, llvm::ArrayRef<long>>(mlir::MLIRContext*, long&&, llvm::ArrayRef<long>&&) (/data/bin/llvm-project/build/bin/mlir-opt+0x43622cb)
 #8 0x000055b040a1c313 mlir::StridedLayoutAttr::get(mlir::MLIRContext*, long, llvm::ArrayRef<long>) (/data/bin/llvm-project/build/bin/mlir-opt+0x4362313)
 #9 0x000055b03eef7954 mlir::memref::SubViewOp::inferResultType(mlir::MemRefType, llvm::ArrayRef<long>, llvm::ArrayRef<long>, llvm::ArrayRef<long>) (/data/bin/llvm-project/build/bin/mlir-opt+0x283d954)
#10 0x000055b03eef7cc9 mlir::memref::SubViewOp::inferResultType(mlir::MemRefType, llvm::ArrayRef<mlir::OpFoldResult>, llvm::ArrayRef<mlir::OpFoldResult>, llvm::ArrayRef<mlir::OpFoldResult>) (/data/bin/llvm-project/build/bin/mlir-opt+0x283dcc9)
#11 0x000055b03ef080f2 SubViewReturnTypeCanonicalizer::operator()(mlir::memref::SubViewOp, llvm::ArrayRef<mlir::OpFoldResult>, llvm::ArrayRef<mlir::OpFoldResult>, llvm::ArrayRef<mlir::OpFoldResult>) (/data/bin/llvm-project/build/bin/mlir-opt+0x284e0f2)
#12 0x000055b03ef087ca mlir::OpWithOffsetSizesAndStridesConstantArgumentFolder<mlir::memref::SubViewOp, SubViewReturnTypeCanonicalizer, SubViewCanonicalizer>::matchAndRewrite(mlir::memref::SubViewOp, mlir::PatternRewriter&) const (/data/bin/llvm-project/build/bin/mlir-opt+0x284e7ca)
#13 0x000055b0435d4c38 mlir::PatternApplicator::matchAndRewrite(mlir::Operation*, mlir::PatternRewriter&, llvm::function_ref<bool (mlir::Pattern const&)>, llvm::function_ref<void (mlir::Pattern const&)>, llvm::function_ref<mlir::LogicalResult (mlir::Pattern const&)>) (/data/bin/llvm-project/build/bin/mlir-opt+0x6f1ac38)
#14 0x000055b040967a8c (anonymous namespace)::GreedyPatternRewriteDriver::processWorklist() GreedyPatternRewriteDriver.cpp:0:0
#15 0x000055b04096befb mlir::applyPatternsAndFoldGreedily(mlir::Region&, mlir::FrozenRewritePatternSet const&, mlir::GreedyRewriteConfig, bool*) (/data/bin/llvm-project/build/bin/mlir-opt+0x42b1efb)
#16 0x000055b0408d11ca (anonymous namespace)::Canonicalizer::runOnOperation() Canonicalizer.cpp:0:0
#17 0x000055b0408b211e mlir::detail::OpToOpPassAdaptor::run(mlir::Pass*, mlir::Operation*, mlir::AnalysisManager, bool, unsigned int) (/data/bin/llvm-project/build/bin/mlir-opt+0x41f811e)
#18 0x000055b0408b25fa mlir::detail::OpToOpPassAdaptor::runPipeline(mlir::OpPassManager&, mlir::Operation*, mlir::AnalysisManager, bool, unsigned int, mlir::PassInstrumentor*, mlir::PassInstrumentation::PipelineParentInfo const*) (/data/bin/llvm-project/build/bin/mlir-opt+0x41f85fa)
#19 0x000055b0408b36c3 mlir::LogicalResult llvm::function_ref<mlir::LogicalResult (mlir::OpPassManager&, mlir::Operation*)>::callback_fn<mlir::detail::OpToOpPassAdaptor::run(mlir::Pass*, mlir::Operation*, mlir::AnalysisManager, bool, unsigned int)::'lambda'(mlir::OpPassManager&, mlir::Operation*)>(long, mlir::OpPassManager&, mlir::Operation*) Pass.cpp:0:0
#20 0x000055b0408e13b3 (anonymous namespace)::InlinerPass::optimizeCallable(mlir::CallGraphNode*, llvm::StringMap<mlir::OpPassManager, llvm::MallocAllocator>&) Inliner.cpp:0:0
#21 0x000055b0408ea2e6 mlir::LogicalResult llvm::function_ref<mlir::LogicalResult ((anonymous namespace)::CallGraphSCC&)>::callback_fn<(anonymous namespace)::InlinerPass::runOnOperation()::'lambda'((anonymous namespace)::CallGraphSCC&)>(long, (anonymous namespace)::CallGraphSCC&) Inliner.cpp:0:0
#22 0x000055b0408eb0d3 (anonymous namespace)::InlinerPass::runOnOperation() Inliner.cpp:0:0
#23 0x000055b0408b211e mlir::detail::OpToOpPassAdaptor::run(mlir::Pass*, mlir::Operation*, mlir::AnalysisManager, bool, unsigned int) (/data/bin/llvm-project/build/bin/mlir-opt+0x41f811e)
#24 0x000055b0408b25fa mlir::detail::OpToOpPassAdaptor::runPipeline(mlir::OpPassManager&, mlir::Operation*, mlir::AnalysisManager, bool, unsigned int, mlir::PassInstrumentor*, mlir::PassInstrumentation::PipelineParentInfo const*) (/data/bin/llvm-project/build/bin/mlir-opt+0x41f85fa)
#25 0x000055b0408b2e24 mlir::PassManager::run(mlir::Operation*) (/data/bin/llvm-project/build/bin/mlir-opt+0x41f8e24)
#26 0x000055b0408a3b93 performActions(llvm::raw_ostream&, std::shared_ptr<llvm::SourceMgr> const&, mlir::MLIRContext*, mlir::MlirOptMainConfig const&) MlirOptMain.cpp:0:0
#27 0x000055b0408a5064 processBuffer(llvm::raw_ostream&, std::unique_ptr<llvm::MemoryBuffer, std::default_delete<llvm::MemoryBuffer>>, mlir::MlirOptMainConfig const&, mlir::DialectRegistry&, llvm::ThreadPool*) MlirOptMain.cpp:0:0
#28 0x000055b0408a5214 mlir::LogicalResult llvm::function_ref<mlir::LogicalResult (std::unique_ptr<llvm::MemoryBuffer, std::default_delete<llvm::MemoryBuffer>>, llvm::raw_ostream&)>::callback_fn<mlir::MlirOptMain(llvm::raw_ostream&, std::unique_ptr<llvm::MemoryBuffer, std::default_delete<llvm::MemoryBuffer>>, mlir::DialectRegistry&, mlir::MlirOptMainConfig const&)::'lambda'(std::unique_ptr<llvm::MemoryBuffer, std::default_delete<llvm::MemoryBuffer>>, llvm::raw_ostream&)>(long, std::unique_ptr<llvm::MemoryBuffer, std::default_delete<llvm::MemoryBuffer>>, llvm::raw_ostream&) MlirOptMain.cpp:0:0
#29 0x000055b0409a4504 mlir::splitAndProcessBuffer(std::unique_ptr<llvm::MemoryBuffer, std::default_delete<llvm::MemoryBuffer>>, llvm::function_ref<mlir::LogicalResult (std::unique_ptr<llvm::MemoryBuffer, std::default_delete<llvm::MemoryBuffer>>, llvm::raw_ostream&)>, llvm::raw_ostream&, bool, bool) (/data/bin/llvm-project/build/bin/mlir-opt+0x42ea504)
#30 0x000055b04089da77 mlir::MlirOptMain(llvm::raw_ostream&, std::unique_ptr<llvm::MemoryBuffer, std::default_delete<llvm::MemoryBuffer>>, mlir::DialectRegistry&, mlir::MlirOptMainConfig const&) (/data/bin/llvm-project/build/bin/mlir-opt+0x41e3a77)
#31 0x000055b0408a5361 mlir::MlirOptMain(int, char**, llvm::StringRef, llvm::StringRef, mlir::DialectRegistry&) (/data/bin/llvm-project/build/bin/mlir-opt+0x41eb361)
#32 0x000055b0408a57c6 mlir::MlirOptMain(int, char**, llvm::StringRef, mlir::DialectRegistry&) (/data/bin/llvm-project/build/bin/mlir-opt+0x41eb7c6)
#33 0x000055b03dbf45db main (/data/bin/llvm-project/build/bin/mlir-opt+0x153a5db)
#34 0x00007f31f3b2f083 __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x24083)
#35 0x000055b03dc8d3ce _start (/data/bin/llvm-project/build/bin/mlir-opt+0x15d33ce)
Aborted (core dumped)
```

### PR DESCRIPTION
**Pull request by rikhuijzer**
This PR fixes https://github.com/llvm/llvm-project/issues/73383 and is another shot at the refactoring proposed in https://github.com/llvm/llvm-project/pull/72885.

### CODE REVIEW

**Review by Lewuathe — COMMENTED** [review]
Is it possible to add the test case (or minimal similar example) with `--inline` option so that we can confirm the original issue is resolved.

https://github.com/llvm/llvm-project/issues/73383

**Comment by rikhuijzer** [comment]
> Is it possible to add the test case (or minimal similar example) with `--inline` option so that we can confirm the original issue is resolved.
> 
> #73383

Thanks for the review and fixing the typos that I've made! I've added a test in a new `MemRef/inlining.mlir` file, which is in line with the other files:
```
$ rg -l '\-inline' mlir/test
mlir/test/Dialect/MemRef/inlining.mlir
mlir/test/Dialect/Vector/inlining.mlir
mlir/test/Dialect/UB/inlining.mlir
mlir/test/Dialect/Async/async-parallel-for-compute-fn.mlir
mlir/test/Dialect/Async/async-parallel-for-canonicalize.mlir
mlir/test/Dialect/Affine/inlining.mlir
mlir/test/Dialect/LLVMIR/inline-byval-huge.mlir
mlir/test/Dialect/Bufferization/inlining.mlir
mlir/test/Dialect/LLVMIR/inlining-alias-scopes.mlir
mlir/test/Dialect/LLVMIR/inlining.mlir
mlir/test/Dialect/Linalg/inline-scalar-operands.mlir
mlir/test/Dialect/Tosa/inlining.mlir
mlir/test/Dialect/Linalg/inlining.mlir
mlir/test/lib/Transforms/TestInlining.cpp
mlir/test/Transforms/inlining.mlir
mlir/test/Transforms/inlining-repeated-use.mlir
mlir/test/Transforms/inlining-recursive.mlir
mlir/test/Transforms/inlining-dce.mlir
mlir/test/Transforms/test-inlining.mlir
```
(Also test additions can always be reverted of course so it shouldn't be too bad.) Also I double checked that https://github.com/llvm/llvm-project/issues/73383 doesn't crash on this PR.

**Review by Lewuathe — APPROVED** [review]
Thanks for the update. It LGTM.

File: mlir/test/Dialect/MemRef/inlining.mlir [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+  cf.br ^bb1(%arg0 : memref<10xf32>)
+^bb1(%m: memref<10xf32>):
+  %c0 = arith.constant 0 : index
+  %c1 = arith.constant 1 : index
+  %1 = memref.subview %m[1] [1] [%c0] : memref<10xf32> to memref<1xf32, strided<[?], offset: 1>>
+  return %1 : memref<1xf32, strided<[?], offset: 1>>
+}
+// CHECK-NEXT: arith.constant 0 : index
+// CHECK-NEXT: %[[SUBVIEW:.*]] = memref.subview
+// CHECK-NEXT: return %[[SUBVIEW]] : memref<1xf32, strided<[?], offset: 1>>
```
Comment (joker-eph) [inline mlir/test/Dialect/MemRef/inlining.mlir:14]: This is a canonicalized crash, we don't need to run the inliner.

**Comment by joker-eph** [comment]
> Is it possible to add the test case (or minimal similar example) with `--inline` option so that we can confirm the original issue is resolved.
> 
> #73383

It is common that the person filing the issue does not reduce the IR or the pipeline that causes the bug. 
In particular this particular reporter (are you associated with them @Lewuathe ?) seems like automated fuzzer. Here the only thing that the inliner does is running the canonicalizer.
I updated the issue instead.

File: mlir/test/Dialect/MemRef/inlining.mlir (hunk 1 above)
Comment (rikhuijzer) [reply mlir/test/Dialect/MemRef/inlining.mlir]: I removed this file again. 👍

**Comment by Lewuathe** [comment]
@joker-eph Thank you. That gets clearer to me!

**Review by joker-eph — APPROVED** [review]
Thanks!

**Pull request merged**